### PR TITLE
Handle logical assignment in super property transforms

### LINKED ISSUE #14163: [Bug]: logical assignment not handled in super property transforms
### 💻

- [X] Would you like to work on a fix?

### How are you using Babel?

Programmatic API (`babel.transform`, `babel.parse`)

### Input code

> Thanks.
> 
> Not related to this PR, but we need to add logical assignment support to super property transforms (they should have been extracted to a new helper):
> 
> ```js
> class A extends B {
>   async foo() {
>     super.foo ??= 1;
>   }
> }
> ```
> 
> [currently throws](https://babeljs.io/repl#?browsers=chrome%2050&build=&builtIns=false&corejs=false&spec=false&loose=false&code_lz=MYGwhgzhAECC0FMAeAXBA7AJjAQtA3gFDTSQCe6w0AZgPa0AUAlAcSdBAK4AOCATgDo6taAH5RAXmgBGANxsAvoQVA&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=script&lineWrap=true&presets=env&prettier=true&targets=&version=7.16.9&externalPlugins=&assumptions=%7B%22noClassCalls%22%3Atrue%2C%22noDocumentAll%22%3Atrue%2C%22mutableTemplateObject%22%3Atrue%2C%22iterableIsArray%22%3Atrue%2C%22ignoreToPrimitiveHint%22%3Atrue%2C%22ignoreFunctionLength%22%3Atrue%2C%22enumerableModuleMeta%22%3Atrue%2C%22noNewArrows%22%3Atrue%2C%22objectRestNoSymbols%22%3Atrue%2C%22privateFieldsAsProperties%22%3Atrue%2C%22pureGetters%22%3Atrue%2C%22setClassMethods%22%3Atrue%2C%22setComputedProperties%22%3Atrue%2C%22setPublicClassFields%22%3Atrue%2C%22setSpreadProperties%22%3Atrue%2C%22skipForOfIteratorClosing%22%3Atrue%2C%22superIsCallableConstructor%22%3Atrue%7D)
> 
> _Originally posted by @JLHwung in https://github.com/babel/babel/pull/14162#pullrequestreview-853689517_

### Configuration file name

_No response_

### Configuration

_No response_

### Current and expected behavior

currently throws error

### Environment

Latest

### Possible solution

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by magic-akari**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | Fixes #14163
| Patch: Bug Fix?          | 👍
| Major: Breaking Change?  |
| Minor: New Feature?      |
| Tests Added + Pass?      | Yes
| Documentation PR Link    | 
| Any Dependency Changes?  |
| License                  | MIT

<a href="https://gitpod.io/#https://github.com/babel/babel/pull/14164"></a>

### CODE REVIEW

File: packages/babel-traverse/test/arrow-transform.js [hunk 1]
```diff
@@ -518,6 +518,152 @@ describe("arrow function conversion", () => {
...
+      };
+      super.foo ??= 4;
+      () => super.foo ??= 4;
+    `,
+      `
+      var _superprop_setFoo = _value => super.foo = _value,
+          _superprop_getFoo = () => super.foo;
+
+      (function () {
+        _superprop_setFoo(_superprop_getFoo() ?? 4);
```
Comment (nicolo-ribaudo) [inline packages/babel-traverse/test/arrow-transform.js:535]: Logical expressions short-circuit, so this should become something like
```js
_superprop_getFoo() ?? _superprop_setFoo(4);
```

Comment (magic-akari) [reply packages/babel-traverse/test/arrow-transform.js:535]: Ah, Thanks

File: packages/babel-traverse/src/path/conversion.ts [hunk 2]
```diff
@@ -397,63 +398,89 @@ function hoistFunctionEnvironment(
...
 
-    const op = assignmentPath.node.operator.slice(0, -1);
+    const op = assignmentPath.node.operator.slice(0, -1) as
+      | LogicalOp
+      | BinaryOp;
+
     const value = assignmentPath.node.right;
 
-    assignmentPath.node.operator = "=";
+    const isLogicalAssignment = op === "||" || op === "&&" || op === "??";
```
Comment (JLHwung) [inline packages/babel-traverse/src/path/conversion.ts:414]: nit: `@babel/types` exports `LOGICAL_OPERATORS` so we don't have to hardcode them.

**Review by JLHwung — APPROVED** [review]
Thanks.

File: packages/babel-traverse/src/path/conversion.ts (hunk 2 above)
Comment (magic-akari) [reply packages/babel-traverse/src/path/conversion.ts]: Updated.
Type narrowing is hard.

**Review by JLHwung — APPROVED** [review]

File: packages/babel-traverse/src/path/conversion.ts [hunk 3]
```diff
@@ -397,63 +399,96 @@ function hoistFunctionEnvironment(
   return { thisBinding, fnPath };
 }
 
-function standardizeSuperProperty(superProp) {
+type LogicalOp = Parameters<typeof logicalExpression>[0];
+type BinaryOp = Parameters<typeof binaryExpression>[0];
+
+function isLogicalOp(op: string): op is LogicalOp {
+  return LOGICAL_OPERATORS.includes(op);
```
Comment (nicolo-ribaudo) [inline packages/babel-traverse/src/path/conversion.ts:406]: If TS allows `const` and `type` with the same name, we could even export `type LOGICAL_OPERATORS = "||" | "&&" | "??"` from `@babel/types`.

**Review by nicolo-ribaudo — APPROVED** [review]
Thanks!

**Pull request merged**